### PR TITLE
Add TLS Security Profile API to NAC

### PR DESCRIPTION
**Pull request by rhrazdil**
**What this PR does / why we need it**:
This PR adds API for configuration of TLSSecurityProfile to NAC custom resource.
TLS configuration is passed via ENV variables to kubemacpool certificate manager pods.

**Special notes for your reviewer**:

**Release note**:

```release-note
Add API for TLSSecurityProfile
```

### CODE REVIEW

File: pkg/network/tlsSecurityProfile.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+func SecurityProfileSpec(profile *ocpv1.TLSSecurityProfile) (string, ocpv1.TLSProtocolVersion) {
+	if profile == nil {
+		return "", ""
+	}
+	if profile.Custom != nil {
+		return strings.Join(profile.Custom.TLSProfileSpec.Ciphers, ","), profile.Custom.TLSProfileSpec.MinTLSVersion
+	}
+	return strings.Join(ocpv1.TLSProfiles[profile.Type].Ciphers, ","), ocpv1.TLSProfiles[profile.Type].MinTLSVersion
+}
```
Comment (qinqon) [inline pkg/network/tlsSecurityProfile.go:17]: Make this functions just select cipher suites and TLS version not converting formats

```suggestion
func SelectCIpherSuitesAndMinTLSVersion(profile *ocpv1.TLSSecurityProfile) ([]string, ocpv1.TLSProtocolVersion) {
	if profile == nil {
		return "", ""
	}
	if profile.Custom != nil {
		return profile.Custom.TLSProfileSpec.Ciphers, profile.Custom.TLSProfileSpec.MinTLSVersion
	}
	return ocpv1.TLSProfiles[profile.Type].Ciphers, ocpv1.TLSProfiles[profile.Type].MinTLSVersion
}
```

File: pkg/network/kubemacpool.go [hunk 2]
```diff
@@ -114,6 +114,10 @@ func renderKubeMacPool(conf *cnao.NetworkAddonsConfigSpec, manifestDir string) (
 	data.Data["CertRotateInterval"] = conf.SelfSignConfiguration.CertRotateInterval
 	data.Data["CertOverlapInterval"] = conf.SelfSignConfiguration.CertOverlapInterval
 
+	ciphers, tlsMinVersion := SecurityProfileSpec(conf.TLSSecurityProfile)
+	data.Data["TLSSecurityProfileCiphers"] = ciphers
```
Comment (qinqon) [inline pkg/network/kubemacpool.go:118]: Here we convert to the format expected by the template
```suggestion
	data.Data["TLSSecurityProfileCiphers"] = strings.Join(ciphers, ",")
```

**Review by qinqon — CHANGES_REQUESTED** [review]

File: pkg/network/tlsSecurityProfile.go (hunk 1 above)
Comment (qinqon) [reply pkg/network/tlsSecurityProfile.go:17]: Also kubemacpool is going to understand the [controller-runtime knob](Kubemacpool is going to understand the controller-runtime knob, can we convert it here ? 

https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/webhook/server.go#L190-L206) can we convert it here ?

File: pkg/components/components.go [hunk 3]
```diff
@@ -799,6 +821,44 @@ func GetCrd() *extv1.CustomResourceDefinition {
 								},
 							},
 						},
+						"tlsSecurityProfile": extv1.JSONSchemaProps{
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply pkg/components/components.go:824]: I wasn't able to use controller-gen, not sure why

File: pkg/components/components.go [hunk 4]
```diff
@@ -585,6 +585,28 @@ func GetCrd() *extv1.CustomResourceDefinition {
...
+			Type: "array",
+		},
+		"minTLSVersion": extv1.JSONSchemaProps{
+			Description: "minTLSVersion is used to specify the minimal version of the TLS protocol that is negotiated during the TLS handshake. For example, to use TLS versions 1.1, 1.2 and 1.3 (yaml): \n   minTLSVersion: TLSv1.1 \n NOTE: currently the highest minTLSVersion allowed is VersionTLS12",
+			Type:        "string",
+			Enum: []extv1.JSON{
+				{Raw: []byte(`"VersionTLS10"`)},
+				{Raw: []byte(`"VersionTLS11"`)},
+				{Raw: []byte(`"VersionTLS12"`)},
+				{Raw: []byte(`"VersionTLS13"`)},
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply pkg/components/components.go]: Done

File: pkg/components/components.go [hunk 5]
```diff
@@ -799,6 +821,44 @@ func GetCrd() *extv1.CustomResourceDefinition {
...
+									Nullable:    true,
+									Type:        "object",
+								},
+								"type": extv1.JSONSchemaProps{
+									Description: "type is one of Old, Intermediate, Modern or Custom. Custom provides the ability to specify individual TLS security profile parameters. Old, Intermediate and Modern are TLS security profiles based on: \n https://wiki.mozilla.org/Security/Server_Side_TLS#Recommended_configurations The profiles are intent based, so they may change over time as new ciphers are developed and existing ciphers are found to be insecure.  Depending on precisely which ciphers are available to a process, the list may be reduced. \n Note that the Modern profile is currently not supported because it is not yet well adopted by common software libraries.",
+									Enum: []extv1.JSON{
+										{Raw: []byte(`"Old"`)},
+										{Raw: []byte(`"Intermediate"`)},
+										{Raw: []byte(`"Modern"`)},
+										{Raw: []byte(`"Custom"`)},
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply pkg/components/components.go]: Done

File: pkg/components/components.go [hunk 6]
```diff
@@ -585,6 +585,28 @@ func GetCrd() *extv1.CustomResourceDefinition {
 		}
 	}
 
+	customSecurityProfileProps := map[string]extv1.JSONSchemaProps{
+		"ciphers": extv1.JSONSchemaProps{
+			Description: "ciphers is used to specify the cipher algorithms that are negotiated during the TLS handshake.  Operators may remove entries their operands do not support.  For example, to use DES-CBC3-SHA  (yaml): \n   ciphers:     - DES-CBC3-SHA",
+			Items: &extv1.JSONSchemaPropsOrArray{
+				Schema: &extv1.JSONSchemaProps{
+					Type: "string",
+				},
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply pkg/components/components.go]: Done, but Enum alone doesn't seem to validate, but it's at least self documenting

File: pkg/network/tlsSecurityProfile.go (hunk 1 above)
Comment (rhrazdil) [reply pkg/network/tlsSecurityProfile.go]: Done

File: pkg/network/kubemacpool.go [hunk 7]
```diff
@@ -114,6 +115,10 @@ func renderKubeMacPool(conf *cnao.NetworkAddonsConfigSpec, manifestDir string) (
 	data.Data["CertRotateInterval"] = conf.SelfSignConfiguration.CertRotateInterval
 	data.Data["CertOverlapInterval"] = conf.SelfSignConfiguration.CertOverlapInterval
 
+	ciphers, tlsMinVersion := SelectCipherSuitesAndMinTLSVersion(conf.TLSSecurityProfile)
+	data.Data["TLSSecurityProfileCiphers"] = strings.Join(ciphers, ",")
+	data.Data["TLSMinVersion"] = tlsMinVersion
```
Comment (qinqon) [inline pkg/network/kubemacpool.go:120]:
```suggestion
	data.Data["TLSMinVersion"] = TLSVersionToHumanReadable(tlsMinVersion)
```

**Review by qinqon — CHANGES_REQUESTED** [review]
Let's do kubemacpool transformations at the same place.

File: pkg/network/tlsSecurityProfile.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+func SelectCipherSuitesAndMinTLSVersion(profile *ocpv1.TLSSecurityProfile) ([]string, string) {
+	if profile == nil {
+		return []string{}, ""
+	}
+	if profile.Custom != nil {
+		return profile.Custom.TLSProfileSpec.Ciphers, TLSVersionToHumanReadable(profile.Custom.TLSProfileSpec.MinTLSVersion)
+	}
+	return ocpv1.TLSProfiles[profile.Type].Ciphers, TLSVersionToHumanReadable(ocpv1.TLSProfiles[profile.Type].MinTLSVersion)
+}
```
[the remark this answers is not in the document]
Comment (rhrazdil) [reply pkg/network/tlsSecurityProfile.go]: Done

File: pkg/network/kubemacpool.go (hunk 7 above)
Comment (rhrazdil) [reply pkg/network/kubemacpool.go]: Done

**Review by qinqon — APPROVED** [review]
/lgtm
/approve

**Pull request merged**